File: exercise-reference/5-claude/solution/tools.ts
import { ABBYYBankMCP } from "./index.js";
import { z } from 'zod';
import { processUtilityBill, UtilityBillData } from './documentAI.js';

export async function initializeTools(agent: ABBYYBankMCP) {

  let billData: UtilityBillData = undefined;

  agent.server.tool("submit-application", "Submit the application to the bank", {}, async () => {
    if (billData === undefined) {
      return {
        content: [{
          type: "text",
          text: "Application cannot be submitted yet. Please upload the required documents."
        }]
      };
    }

    return {
      content: [{
        type: "text",
        text: "Application submitted successfully to ABBYY bank. We'll be in touch shortly after reviewing your application."
      }]
    };
  });

  agent.server.tool(
    "upload-utility-bill",
    "Upload and process an image of a utility bill",
    {
      documentFilepath: z.string().describe("The path to the document file to process")
    },
    async ({ documentFilepath }: { documentFilepath: string }) => {
      try {
        billData = await processUtilityBill(documentFilepath);
        
        return {
          content: [{
            type: "text",
            text: JSON.stringify({
              status: "success",
              message: "Utility bill processed successfully",
              data: billData
            }, null, 2)
          }]
        };
      } catch (error) {
        console.error("Error processing utility bill:", error);
        return {
          isError: true,
          content: [{
            type: "text",
            text: `Error processing utility bill: ${error instanceof Error ? error.message : String(error)}`
          }]
        };
      }
    }
  );
}

File: py/exercise-reference/3-docai/tools.py
from document_ai import process_utility_bill

def register_tools(mcp, state):
    @mcp.tool()
    async def upload_utility_bill(documentFilepath: str) -> dict:
        try:
            bill_data = await process_utility_bill(documentFilepath)
            
            # Store the processed data
            state["documents"]["utilityBill"] = bill_data
            
            return {
                "content": [{
                    "type": "text",
                    "text": f"Utility bill processed successfully\n\nCustomer Information:\n- Full Name: {bill_data.full_name}\n- Account Number: {bill_data.account_number}\n- Address: {bill_data.address['street']}\n- Issue Date: {bill_data.issue_date}\n- Utility Provider: {bill_data.utility_provider}"
                }]
            }
        except Exception as error:
            print(f"Error processing utility bill: {error}")
            return {
                "isError": True,
                "content": [{
                    "type": "text",
                    "text": f"Error processing utility bill: {str(error)}"
                }]
            } 
